File: solutions/Golang/valid_sudoku_hash_sets.go
func isValidSudoku(board [][]byte) bool {
    var rows, cols, boxes [9]map[byte]bool
    for i := 0; i < 9; i++ {
        rows[i]  = map[byte]bool{}
        cols[i]  = map[byte]bool{}
        boxes[i] = map[byte]bool{}
    }

    for r := 0; r < 9; r++ {
        for c := 0; c < 9; c++ {
            d := board[r][c]
            if d == '.' {
                continue
            }
            b := (r/3)*3 + (c / 3)
            if rows[r][d] || cols[c][d] || boxes[b][d] {
                return false
            }
            rows[r][d] = true
            cols[c][d] = true
            boxes[b][d] = true
        }
    }
    return true
}

File: solutions/Rust/valid_sudoku_hash_sets.rs
use std::collections::HashSet;

impl Solution {
    pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {
        let mut rows:  [HashSet<char>; 9] = Default::default();
        let mut cols:  [HashSet<char>; 9] = Default::default();
        let mut boxes: [HashSet<char>; 9] = Default::default();

        for r in 0..9 {
            for c in 0..9 {
                let d = board[r][c];
                if d == '.' { continue; }
                let b = (r / 3) * 3 + (c / 3);
                if !rows[r].insert(d) || !cols[c].insert(d) || !boxes[b].insert(d) {
                    return false;
                }
            }
        }
        true
    }
}
